src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp

```diff
@@ -498,6 +501,262 @@ void runtime::gpu::CudaKernelBuilder::get_softmax_op(codegen::CodeWriter& writer
     return;
 }
 
+void runtime::gpu::CudaKernelBuilder::get_softmax_block_reduce_op(
+    codegen::CodeWriter& writer,
+    const std::string& name,
+    runtime::gpu::GPUKernelArgs& args,
+    const std::vector<std::string>& data_types,
+    size_t non_reduce_rank,
+    size_t reduce_rank,
+    size_t block_size_x)
+{
+    writer << runtime::gpu::nvrtc::helpers();
```

Comment:
nit: good opportunity for read-only loads since the same input addresses are loaded from several times
```suggestion
    writer << runtime::gpu::nvrtc::helpers();
    writer << runtime::gpu::nvrtc::define_non_coherent_load(data_types[1], "load");
```

Reply:
seems it's a little bit slower in this case.  `3,7xx us vs 3,5xx us`.

Reply:
hmm, Ok, with more test, it's about the same on average. I guess it's because here data is been kicked out of cache anyway before the next read. So it does not help. But not hurt too. I might here is data size is smaller for each warp.

Reply:
keep the load here.

Reply:
Makes sense. I had tested this previously in the nd case where reduce_stride !=1 (the `::get_softmax_op` impl. in the previous softmax PR) and it gave a ~20% perf. boost. That was using a smaller tensor from an rn50 json.